clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,506 @@
+func (npm *networkPerformanceMeasurement) getWeightedPodReadyTimeout() time.Duration {
+	weightedPodReadyTimeout := float32(npm.numberOfServers+npm.numberOfClients) * podReadyTimeOutFactor
+	podReadyTimeout := time.Duration(weightedPodReadyTimeout) * time.Second
+
+	klog.Info("podReadyTimeout: ", podReadyTimeout)
+	return podReadyTimeout
+}
+
+func (*networkPerformanceMeasurement) String() string {
+	return networkPerformanceMetricsName
+}
+
+func (npm *networkPerformanceMeasurement) storeWorkerPods() error {
+	var podsWithIPs int
+	options := metav1.ListOptions{}
+	pods, err := npm.k8sClient.CoreV1().Pods(netperfNamespace).List(context.TODO(), options)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods.Items {
+		if pod.Status.PodIP != "" {
+			podData := &workerPodData{podName: pod.Name, podIP: pod.Status.PodIP, workerNode: pod.Spec.NodeName}
+			npm.populateWorkerPodList(podData)
+			podsWithIPs++
+		}
+	}
+	klog.Infof("Pods configured: %v, Pods with IPs: %v", npm.numberOfServers+npm.numberOfClients, podsWithIPs)
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) gather() (measurement.Summary, error) {
+	npm.createResultSummary()
+	content, err := util.PrettyPrintJSON(&measurementutil.PerfData{
+		Version:   "v1",
+		DataItems: npm.resultSummary.DataItems,
+	})
+	if err != nil {
+		klog.Info("Pretty Print to Json Err: ", err)
+	}
+	summaryName := fmt.Sprintf(npm.String() + npm.resultSummary.podRatio + npm.resultSummary.protocol + npm.resultSummary.service)
+	return measurement.CreateSummary(summaryName, "json", content), nil
+}
+
+// Dispose disposes resources,objects after the measurement.
+func (npm *networkPerformanceMeasurement) Dispose() {
+	if npm.framework == nil {
+		klog.V(1).Infof("Network measurement %s wasn't started, skipping the Dispose() step", npm)
+		return
+	}
+	klog.Infof("Stopping network performance measurement...")
+	close(npm.stopCh)
+	close(npm.podInfo.podPairCh)
+	npm.cleanupCluster()
+}
+
+func (npm *networkPerformanceMeasurement) handleUpdateNetworkTestEvents(newObj interface{}) {
+	newRuntimeObj, ok := newObj.(runtime.Object)
+	if newObj != nil && !ok {
+		klog.Errorf("error casting object: %s", newObj)
+		return
+	}
+
+	resourceContent, err := runtime.DefaultUnstructuredConverter.ToUnstructured(newRuntimeObj)
+	if err != nil {
+		klog.Errorf("error converting event Object to unstructured.Event object: %s", newObj)
+		return
+	}
+	var metricResp metricResponse
+	metricMap := resourceContent["status"].(map[string]interface{})
+	if err = getMetricResponse(metricMap, &metricResp); err != nil {
+		klog.Error(err)
+	}
+	if len(metricResp.Metrics) == 0 && metricResp.Error != "" {
+		klog.Error("error message from worker pod: ", metricResp.Error)
+	}
+	npm.metricLock.Lock()
+	npm.metricVal = append(npm.metricVal, metricResp.Metrics)
+	npm.metricLock.Unlock()
+}
+
+func (npm *networkPerformanceMeasurement) execNToMTest() {
+	go npm.createCustomResourcePerUniquePodPair()
```

Comment:
I don't understand why you want to do that asynchronous.

Why not simply:
- first create all of those pairs and store them in some slice
- then just iterate over this slice and create appropriate  CRs ?

Reply:
When number of pods are higher , we thought going the synchronous way , we be slower in comparison to current implementation.

Reply:
These methods are so simple that I'm completely against introducing unnecessary complexity.
You won't have millions of pods, and for hundreds or even small thousands it doesn't matter.